ingest-app/src/cmr/ingest/data/granule_bulk_update.clj

```diff
@@ -0,0 +1,253 @@
+ (update-bulk-update-granule-status [db task-id concept-id status status-message])
+ (reset-bulk-update [db]))
+
+(def bulk_update_task_status
+  "bulk_update_gran_status")
+
+;; Extends the BulkUpdateStore to the oracle store so it will work with oracle.
+(extend-protocol BulkUpdateStore
+ cmr.oracle.connection.OracleStore
+
+ (get-provider-bulk-update-status
+   [db provider-id]
+   (j/with-db-transaction
+     [conn db]
+     ;; Returns a list of bulk update statuses for the provider
+     (let [stmt (su/build (su/select [:created-at :name :task-id :status :status-message :instruction]
+                                     (su/from bulk_update_task_status)
+                                     (su/where `(= :provider-id ~provider-id))))
+           ;; Note: the column selected out of the database is created_at, instead of created-at.
+           statuses (doall (map #(update % :created_at (partial oracle/oracle-timestamp->str-time conn))
+                                (su/query conn stmt)))
+           statuses (map util/map-keys->kebab-case statuses)]
+       (map #(update % :request-json-body util/gzip-blob->string)
+            statuses))))
+
+ (get-bulk-update-task-status
+   [db task-id provider-id]
+   (j/with-db-transaction
+     [conn db]
+     ;; Returns a status for the particular task
+     (some-> conn
+             (su/find-one (su/select [:created-at :name :status :status-message :instruction]
+                                     (su/from bulk_update_task_status)
+                                     (su/where `(and (= :task-id ~task-id)
+                                                     (= :provider-id ~provider-id)))))
+             util/map-keys->kebab-case
+             (update :request-json-body util/gzip-blob->string)
+             (update :created-at (partial oracle/oracle-timestamp->str-time conn)))))
+
+ (get-bulk-update-task-granule-status
+   [db task-id]
+   ;; Get statuses for all granules by task id
+   (map util/map-keys->kebab-case
+        (su/query db (su/build (su/select [:granule-ur :status :status-message]
+                                          (su/from "bulk_update_gran_status")
+                                          (su/where `(= :task-id ~task-id)))))))
+
+ (get-bulk-update-granule-status
+   [db task-id granule-ur]
+   ;; Get the status for a particular granule
+   (su/find-one db (su/select [:status :status-message]
+                              (su/from "granule_bulk_update_tasks")
+                              (su/where `(and (= :task-id ~task-id)
+                                              (= :granule-ur ~granule-ur))))))
+
+ (create-and-save-bulk-update-status
+   [db provider-id json-body granule-urs]
+   ;; In a transaction, add one row to the task status table and for each concept
+   (j/with-db-transaction
+     [conn db]
+     (let [task-id (:nextval (first (su/query db ["SELECT task_id_seq.NEXTVAL FROM DUAL"])))
```

Comment:
how likely is it that another request will grab the same task-id?

Reply:
This is the same logic that's being used in collection bulk update, so hopefully not very likely 😅

Reply:
I yield to precedence, but I don't like it.